### PR TITLE
Opt-in behaviour to avoid name pollution

### PR DESCRIPTION
**Pull request by shadowspawn**
# Pull Request

## Problem

Commander has traditionally stored the option values on the command object. This gives a compact syntax especially when working with the program object, but means there are potential naming conflicts between user defined options and properties on Command. This catches people out and there are not good work-arounds.

Collected issues: #933

For example, code with a conflict on both program and a command::

```
const commander = require('commander');

const program = new commander.Command();
program
  .option('--name <value>', 'specify name');
program
  .command('sub')
  .option('--action <value>', 'specify action')
  .action((options) => {
    console.log(options.action);
  });

program.parse(process.argv);

options = program.opts();
console.log(options.name);
```

```
% node pollution.js sub
[Function]
[Function]
% node pollution.js --name NAME sub --action ACTION
ACTION
NAME
```

## Solution

Add a way of configuring Command to behave differently. Implement "modern" configuration which stores the option values separately, and passes just the options to action handlers. The main code change is accessing program options through (existing) `.opts()` member.

- backwards compatible. Existing behaviour unchanged.
- often no changes required in action handlers. Most people probably think of the parameter as for just the options anyway!
- configurable for special cases or existing advanced usage requiring command object passed to action handler

With the addition of couple of lines to the example program above, no other code changes:

```
...
const program = new commander.Command();
program
  .storeOptionsAsProperties(false)
  .passCommandToAction(false);
...
```

Now get:

```
% node pollution.js sub                            
undefined
undefined
% node pollution.js --name NAME sub --action ACTION
ACTION
NAME
```

## ToDo

Add info to README.

## Notes

I considered using a Map for the option values rather than a plain Object, but it would mean changing all existing client code accessing option values. Continuing to use an object means action handling code is largely unaffected.

I considered creating option values object with `Object.create(null)` to further reduce name collisions, but felt it was too unusual for the small benefits. 

I would have exposed a data property called `opts` if there was not already a function with that name! However, using the existing function does allow code that works with both the old way and the new way, which is nice.

We could change all the code examples to use `program.opts()` so they work with both old and modern behaviour.

In a future major version we could change the default behaviour. The routines are named for the future, where you need to opt back into the old behaviour.

### CODE REVIEW

**Comment by shadowspawn** [comment]
ts-node test still unreliable.

Previously noted: https://github.com/tj/commander.js/pull/1095#issuecomment-552376208

**Comment by shadowspawn** [comment]
I am rethinking "modern" name as not suitable for the longterm, and considering other names for this combo configuration.

**Comment by shadowspawn** [comment]
My current idea is a named combo like `{ combo: 'safeOptions' }` instead of `{ modern: true }`. This is clearer that it is just a combination of  other settings and not something new, and allows other combinations in the future especially for backwards compatibility when we change default behaviour, like say `{ combo: 'v2Compatible' }`.

**Comment by shadowspawn** [comment]
Still not happy with the interface using `configureCommand`. Will experiment with routines in more typical Commander style, like `. storeOptionsAsProperties(false)`.

**Comment by shadowspawn** [comment]
Old behaviour:
```
const program = new commander.Command();
```

New behaviour, safe options, which are passed to action handler:
```
const program = new commander.Command();
program
   .storeOptionsAsProperties(false)
   .passCommandToAction(false);
```

(The routines are named for the future when the default behaviour has changed, and would pass true to get the legacy behaviour.)

**Comment by shadowspawn** [comment]
Added section in README covering new routines and rationale, and example files for the three expected code patterns.

Ready for feedback/review.

**Comment by shadowspawn** [comment]
Woo hoo! 🎉

File: index.js [hunk 1]
```diff
@@ -531,6 +540,70 @@ Command.prototype.allowUnknownOption = function(arg) {
...
+ *
+ * @param {String} key
+ * @return {Object} value
+ * @api private
+ */
+
+Command.prototype._getOptionValue = function(key) {
+  if (this._storeOptionsAsProperties) {
+    return this[key];
+  } else {
```
Comment (abetomo) [inline index.js]: This is a personal preference, but this 'else' can be deleted.

**Comment by shadowspawn** [comment]
I am thinking perhaps do next release in early January?

(I am starting to get busy leading into Xmas so my time will be limited.)

File: index.js (hunk 1 above)
Comment (shadowspawn) [reply index.js]: Short version: I'll look into this a little, and probably delete the else.

Long version: I do not use an else when it is an early return like for an error, but I sometimes keep the if/else when they are both normal situations. However, I know some people prefer not to have the redundant else and am ok with that. I think I saw a lint option for this recently. I'll have a look and see if I am in the minority in this code, and we could enforce it. :-)

Comment (shadowspawn) [reply index.js]: There were only two cases, and they were both introduced by me in this pull request! I have changed them both and added a lint rule.

Comment (abetomo) [reply index.js]: Thank you!

Comment (shadowspawn) [reply index.js]: Thanks-you for the suggestion. 🙂

**Comment by shadowspawn** [comment]
Shipped in Commander v4.1

**Pull request merged**